enterprise/server/backends/pebble_cache/pebble_cache.go

```diff
@@ -566,6 +629,110 @@ func (p *PebbleCache) updateAtime(key filestore.PebbleKey) error {
 	return db.Set(keyBytes, protoBytes, &pebble.WriteOptions{Sync: false})
 }
 
+func (p *PebbleCache) migrateData(quitChan chan struct{}) error {
+	if len(p.migrators) == 0 {
+		log.Debugf("No migrations necessary")
+		return nil
+	}
+
+	limiter := rate.NewLimiter(rate.Limit(*migrationQPSLimit), 1)
+	db, err := p.leaser.DB()
+	if err != nil {
+		return err
+	}
+	defer db.Close()
+
+	iter := db.NewIter(&pebble.IterOptions{
+		LowerBound: keys.MinByte,
+		UpperBound: keys.MaxByte,
+	})
+	defer iter.Close()
+
+	minVersion := p.maxDatabaseVersion()
+	maxVersion := p.minDatabaseVersion()
+	migrationStart := time.Now()
+	keysSeen := 0
+	keysMigrated := 0
+	lastStatusUpdate := time.Now()
+
+	for iter.First(); iter.Valid(); iter.Next() {
+		if bytes.HasPrefix(iter.Key(), SystemKeyPrefix) {
+			continue
+		}
+		keysSeen += 1
+
+		select {
+		case <-quitChan:
+			return nil
+		default:
+		}
+		_ = limiter.Wait(p.env.GetServerContext())
+
+		if time.Since(lastStatusUpdate) > 10*time.Second {
+			log.Infof("Pebble Cache: data migration progress: saw %d keys, migrated %d to version: %d in %s", keysSeen, keysMigrated, maxVersion, time.Since(migrationStart))
+			lastStatusUpdate = time.Now()
+		}
+		var key filestore.PebbleKey
+		version, err := key.FromBytes(iter.Key())
+		if err != nil {
+			return err
+		}
+		oldVersion := version
+		valBytes := iter.Value()
+
+		for _, migrator := range p.migrators {
+			// If this key was already migrated, skip it.
+			if version >= migrator.ToVersion() {
+				continue
+			}
+
+			// If this key does not match this migrator's
+			// "FromVersion", then the migrators that should have
+			// run before this one did not, and something is wrong.
+			// Bail out.
+			if version != migrator.FromVersion() {
+				return status.FailedPreconditionErrorf("Migrator %+v cannot migrate key from version %d", migrator, version)
+			}
+
+			valBytes = migrator.Migrate(valBytes)
+			version = migrator.ToVersion()
+			keysMigrated += 1
+
+			if version > maxVersion {
+				maxVersion = version
+			}
+			if version < minVersion {
+				minVersion = version
+			}
+
+		}
+		if version == oldVersion {
+			continue
+		}
+
+		keyBytes, err := key.Bytes(version)
```

Comment:
I think you need to lock the key here. Someone could do a read-then-modify operation on the key you are migrating and unintentionally restore it.